traffic_ops/traffic_ops_golang/topology/queue_update.go

```diff
@@ -0,0 +1,100 @@
+	"github.com/apache/trafficcontrol/lib/go-util"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/dbhelpers"
+	validation "github.com/go-ozzo/ozzo-validation"
+	"net/http"
+)
+
+func Validate(reqObj tc.TopologiesQueueUpdateRequest, topologyName tc.TopologyName, tx *sql.Tx) error {
+	errorMap := validation.Errors{}
+
+	if reqObj.Action != "queue" && reqObj.Action != "dequeue" {
+		errorMap["action"] = errors.New("action must be 'queue' or 'dequeue'")
+	}
+
+	if _, cdnExists, err := dbhelpers.GetCDNNameFromID(tx, reqObj.CDNID); err != nil {
+		errorMap["cdnId"] = fmt.Errorf("could not check whether cdn exists for id %d", reqObj.CDNID)
+	} else if !cdnExists {
+		errorMap["cdnId"] = fmt.Errorf("no cdn exists with id %d", reqObj.CDNID)
+	}
+
+	if topologyExists, err := dbhelpers.TopologyExists(tx, topologyName); err != nil {
+		errorMap["topology"] = fmt.Errorf("could not check whether topology %s exists", topologyName)
+	} else if !topologyExists {
+		errorMap["topology"] = fmt.Errorf("no topology exists by the name of %s", topologyName)
+	}
+
+	return util.JoinErrs(tovalidate.ToErrors(errorMap))
+}
+
+func QueueUpdateHandler(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, []string{"name"}, []string{})
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+	reqObj := tc.TopologiesQueueUpdateRequest{}
+	if err := json.NewDecoder(r.Body).Decode(&reqObj); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("malformed JSON: "+err.Error()), nil)
+		return
+	}
+	topologyName := tc.TopologyName(inf.Params["name"])
+	if err := Validate(reqObj, topologyName, inf.Tx.Tx); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, fmt.Errorf("invalid request to queue updates: %s", err), nil)
+		return
+	}
+	if err := queueUpdates(inf.Tx.Tx, topologyName, reqObj.CDNID, reqObj.Action == "queue"); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("Topology queueing updates: "+err.Error()))
+		return
+	}
+
+	message := fmt.Sprintf("TOPOLOGY: %s, ACTION: Topology server updates %sd", topologyName, reqObj.Action)
+	api.CreateChangeLogRawTx(api.ApiChange, message, inf.User, inf.Tx.Tx)
+	api.WriteResp(w, r, tc.TopologiesQueueUpdate{Action: reqObj.Action, CDNID: reqObj.CDNID, Topology: topologyName})
+}
+
+func queueUpdates(tx *sql.Tx, topologyName tc.TopologyName, cdnId int64, queue bool) error {
+	query := `
+UPDATE server s
+SET upd_pending = $1
+FROM cachegroup c, topology_cachegroup tc, cdn
+WHERE s.cachegroup = c.id
+AND c."name" = tc.cachegroup
+AND tc.topology = $2
+AND s.cdn_id = $3
+`
+	var err error
+	if _, err = tx.Exec(query, queue, topologyName, cdnId); err != nil {
+		err = fmt.Errorf("queueing updates: %s", err)
```

Comment:
Should this be `err.Error()` instead?

Reply:
*Should* it, as in will it not work unless it is `err.Error()`? `fmt`'s reflection successfully casts `Error` to `string` type because it only has 1 field, which is `string` type, so `%s` works.